### PR TITLE
CA bundle mount options for storage initializer

### LINKED ISSUE #2897: Custom certificate bundle for storage initializer
/kind feature

**Describe the solution you'd like**

The current version of KServe provides the `serving.kserve.io/s3-cabundle` annotation that allows administrators to specify the path to a certificate bundle for HTTPS validations during S3 resource initialization. However, I'm wondering if there is a way to mount a certificate bundle in the storage initializer container without having to rebuild the image.

Can KServe mount volume from secret into the container it generates? This would allow administrators to easily mount the necessary certificate bundle without the need to rebuild the container image.

**Anything else you would like to add:**

Example based on [Knative](https://knative.dev/docs/serving/tag-resolution/#procedure):

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: controller
  namespace: knative-serving
spec:
  template:
    spec:
      containers:
        - name: controller
          volumeMounts:
            - name: custom-certs
              mountPath: /path/to/custom/certs
          env:
            - name: SSL_CERT_DIR
              value: /path/to/custom/certs
      volumes:
        - name: custom-certs
          secret:
            secretName: custom-certs
```

### PR DESCRIPTION
**Pull request by danilopeixoto**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2897

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Test injected CA bundle secret volume
- [x] Test not injected CA bundle secret volume

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
CA bundle mount options for storage initializer
```

### CODE REVIEW

File: config/configmap/inferenceservice.yaml [hunk 1]
```diff
@@ -23,6 +23,8 @@ data:
         "cpuRequest": "100m",
         "cpuLimit": "1",
         "storageSpecSecretName": "storage-config",
+        "caBundleSecretName": "",
```
Comment (yuzisun) [inline config/configmap/inferenceservice.yaml:420]: @danilopeixoto I guess this secret is required to be created for each namespace where the inference service pod is running right?

**Comment by Jooho** [comment]
@danilopeixoto @yuzisun  This is a great idea and I also look for the same thing. In my opinion, this functionality should be supported using annotations at the inference service level. The reason is that not all inference services necessarily need to use the same CA bundle. Therefore, there is no need to create unnecessary cabundle secrets in all Kserve namespaces. By utilizing annotations, each inference service can specify its own required CA bundle, making the configuration more flexible and efficient. What do you think?

**Comment by Jooho** [comment]
@yuzisun @danilopeixoto does this PR move forward?

**Comment by yuzisun** [comment]
> @yuzisun @danilopeixoto does this PR move forward?

Yes sorry for the delay, meanwhile can you help resolve the conflicts ?

**Comment by Jooho** [comment]
@yuzisun I can definitely help this PR but I am not sure how I can add a new commit to solve this conflicts.
Do I need to send a new PR based on @danilopeixoto's PR ?

**Comment by yuzisun** [comment]
> @yuzisun I can definitely help this PR but I am not sure how I can add a new commit to solve this conflicts. Do I need to send a new PR based on @danilopeixoto's PR ?

@Jooho I resolved the conflicts, could you help test the PR?

**Comment by Jooho** [comment]
@yuzisun sorry for the late response. 
I think we need to extend this pr more. This is my idea regarding ca bundle support. We need 2 options : global, isvc specific. If users set the cabundle secret to global configuration configMap, kserve operator needs to create the secret (ca bundle) per each user namespace. but if users want to use their own secret(cabundle) in the namespace, it needs to be possible to set the secret(ca bundle) with an inferenceService.
If there is a specific annotation specified in an inferenceServince, kserve operator should not create a global secret in a namespace and use the secret user specified in the user namespace. 
@yuzisun @ckadner What do you think?
this is a very high-level diagram.
image

**Comment by Jooho** [comment]
We can work separately. The first global option then isvc annotation.

By the way, this pr missed creating a secret in each namespace. This needs to be added.

**Comment by yuzisun** [comment]
> We can work separately. The first global option then isvc annotation.
> 
> By the way, this pr missed creating a secret in each namespace. This needs to be added.

Thanks for the explanation! I think it makes sense, we can limit this PR to be the global scoped CA secret and create a separate PR for implementing the namespace level CA bundle annotation. I have a question though, how do you plan to provision the CA secrets to the user namespaces ? we do not know which namespaces are kserve user namespaces.

**Review by ckadner — CHANGES_REQUESTED** [review]
I created a development branch `ca_bundle_colab` and a corresponding PR #3216 so @danilopeixoto and @Jooho can work together to complete the work started in this PR

I'm leaving this PR open for now, but set it into draft mode. Once the work is completed on PR #3216, with @danilopeixoto and @Jooho as co-authors, this PR should be closed.

**Comment by ckadner** [comment]
> I created a development branch `ca_bundle_colab` and a corresponding PR #3216 so @danilopeixoto and @Jooho can work together to complete the work started in this PR
> 
> I'm leaving this PR open for now, but set it into draft mode. Once the work is completed on PR #3216, with @danilopeixoto and @Jooho as co-authors, this PR should be closed.

The collaboration branch idea did not pan out. @Jooho just created a new PR with all of @danilopeixoto commits to establish co-authorship

https://github.com/kserve/kserve/pull/3250

**Pull request closed without merging**